### DISCUSSION TITLE
Sync comments from focalboard

### DISCUSSION THREAD

**Discussion by michelengelen** [mattermost/community#79] [Practical Ideas]
The other day I had this idea, when @zefhemel and I were discussing in messaging and focalboard synchronously, that it would be a great feature to sync the comments from focalboard items when you paste a link to it in the channel.

I imagine it as a preview-like post, that has a thread going where all current/existing comments get ported to the thread and you get the ability to reply to it. Any replies written will then get synced with the focalboard item as comments as well.

WDYT @chenilim and @esethna ?

**Comment by chenilim** [comment]
Thanks @michelengelen! Yes, we want to do this. Here's [the card tracking this](https://focalboard-community.octo.mattermost.com/workspace/qgsck6cts3fwpqwyjiupjm5cde/shared?id=47aa9bb4-6967-4a96-83c7-11bd6b20f1eb&v=ca1a5441-4c1c-4d0e-856e-88cc744576c8&r=762db9c6-305c-4d13-8ce6-9f11e5232c24&c=f90dde37-8188-4bdf-881e-cdbbaaf29044). Current thinking is it would be great if we could simply make comments on a Focalboard card (Focalcard?) a comment thread. So you can either comment on the card, or reply to the thread in Messaging. Sounds in line with your description above.

Lots of details to work out. One way perhaps is:
* When you make the first comment, it creates a new thread on the associated channel
* And stores the thread Id with the card
* Subsequent comments just reply to that thread
* And the comment display is simply a query on the Messaging API

cc @michaelgamble, @sbishel.

**Reply by michelengelen** [reply]
great ideas ... I can see it become overloaded with focalcards though on very populated channels with high traffic. What do you think about adding a RHS panel to it that collects all focalcards that were shared with that channel?

**Reply by chenilim** [reply]
Good question! Short answer is I think [collapsed reply threads](https://mattermost.com/blog/dev-sneak-peek-collapsed-reply-threads/#:~:text=Reply%20messages%20are%20collapsed%20under,in%20the%20right%2Dhand%20side) is the general solution to that problem.
* The RHS idea is worth exploring
* And / or maybe an extension to collapsed reply threads that have a special display when then thread is about a Focalcard.

However, you raise an additional important nuance: How do we track multiple threads per card? i.e. if I share a link to a Focalcard on 5 different channels, is there a way to track and view those 5 threads in one place?

This is likely a few steps out, but my dream scenario would be that all threads that mention or link to a card are shown on the card itself. Analogous to backlinks, like how mentioning a GitHub issue (like this: #82) creates a backlink in that issue as well. Extra dream scenario is we can reference cards in Messaging the same way, e.g. `#card-1234` one day.

**Reply by michelengelen** [reply]
I think the API should be handled by focalboard. I dont see a problem sharing a card in multiple channels where all comments are in sync... it would actually be a good way of cross-channel collaboration, don't you think? 👍🏻

**Reply by chenilim** [reply]
Ok, that's an even more mind-expanding idea, sharing a thread across multiple channels. :) My initial instinct is different people or groups will want to have separate conversations around a card. E.g. for a feature card, there's a team status conversation, but could also be a 1:1 conversation when discussing individual work queues.

I agree there generally is one "main" or "official" comment thread on a card. To your original point, there could be some preview of this when pasting a link in a channel. We need to think about how the user differentiates between adding to the "official" thread, vs. commenting on a secondary thread in the channel.

**Reply by michelengelen** [reply]
That's pretty interesting. One solution could be to restrict access to certain users, or make it open to all. 

That way you could also restrict "sharability" in channels, since the users are shared among messaging and focalboard.

Take the 1:1 example: the manager creates a focalcard and restricts access to himself and the team-member regarding the 1:1. with this setting you could also enable a restriction where the card could be shared (only in DM channel with his team-member).

This could work the same way in GM channels as well.

There is only one question: how to handle a change in this setting (change users on the card, or e.g. adding a user without access rights to a GM channel).

Any ideas?

**Reply by chenilim** [reply]
Ok, now we are wading into the deep end of the pool. :) This is at the heart of what we are grappling with - to offer granular permissions without being over-complicated for the user. No easy answers, but one way we can simplify is to cleanly separate "public" (to the team, or anyone with a login) and "private" boards.

**Public boards** are expected to be freely shared anywhere, i.e pasted into multiple channels, commented on by anyone, even (if publicly writable) edited by anyone. Analogous to a Google Doc that's shared with anyone in the org.

**Private boards** are restricted to a specific set of users (or groups). Perhaps we show a warning when pasting links to them (or their cards) in channels ("only specific users have access"). Anyone can see the link, but only authorized users can see the preview or open the item.

Right now, I think the baseline is permissions like Google Docs. I don't think we need a "shareability" restriction on top of that - anyone can see the link, but you need permissions to view the content. Still, lots to think about.

**Reply by michelengelen** [reply]
ok, i see where this is headed! :D

It sounds like a lot of thought is already gone into this. All of this leaves me with just one last question: If anyone can see the link will only the users with permission be able to see a focalcard-preview?

As I imagine it the permission stuff should be handled a bit differently: don't show a warning to the users after pasting it, but show one to the user who tries to paste while doing so.

`Warning: Only users with the permission to view this focalcard will be able to see and collaborate on it. User without permission: @john.doe, ...`

Maybe it would be a great feature to give the users a opportunity to extend permissions to those that do not currently have it:

`You can grant permission to the user by "XYZ"`

WDYT?

**Reply by chenilim** [reply]
Yeah, I was thinking the same as you (show the warning to the user doing the pasting). In general, this is the same problem when pasting links to things like Google Docs. The preview should only be visible to authorized users (with read permissions).

Google has a "request permissions" UX that we could follow.

Have to think more, but sharing a link to a restricted Focalboard is like sharing a link to a thread in a private channel. It's not clear if the user's intent is to **expand access** or to only address the "intersection" of users with permissions. Making it easy to grant additional permissions would be nice.

**Reply by michaelgamble** [reply]
This model of focalboard link preview permissions could follow the same permissions model of permalink previews (this feature just went in) effectively controlling so only members of a private channel will see the preview of the permalink post ... Anyone without access would only see the link... In theory we can reuse this
